ledger/consumer/smartcontract/tx.go

```diff
@@ -0,0 +1,412 @@
+	}
+}
+
+// NewSpawn returns a new spawn transaction.
+func (f TransactionFactory) NewSpawn(contractID string,
+	arg proto.Message) (SpawnTransaction, error) {
+
+	if arg == nil {
+		return SpawnTransaction{}, xerrors.New("argument cannot be nil")
+	}
+
+	tx := SpawnTransaction{
+		encoder:    f.encoder,
+		ContractID: contractID,
+		Argument:   arg,
+	}
+
+	hash, err := tx.computeHash(f.hashFactory, f.marshaler)
+	if err != nil {
+		return tx, xerrors.Errorf("couldn't hash tx: %v", err)
+	}
+
+	tx.transaction = transaction{hash: hash}
+
+	return tx, nil
+}
+
+// NewInvoke returns a new invoke transaction.
+func (f TransactionFactory) NewInvoke(key []byte, arg proto.Message) (InvokeTransaction, error) {
+	if arg == nil {
+		return InvokeTransaction{}, xerrors.New("argument cannot be nil")
+	}
+
+	tx := InvokeTransaction{
+		encoder:  f.encoder,
+		Key:      key,
+		Argument: arg,
+	}
+
+	hash, err := tx.computeHash(f.hashFactory, f.marshaler)
+	if err != nil {
+		return tx, xerrors.Errorf("couldn't hash tx: %v", err)
+	}
+
+	tx.transaction = transaction{hash: hash}
+
+	return tx, nil
+}
+
+// NewDelete returns a new delete transaction.
+func (f TransactionFactory) NewDelete(key []byte) (DeleteTransaction, error) {
+	tx := DeleteTransaction{
+		Key: key,
+	}
+
+	hash, err := tx.computeHash(f.hashFactory)
+	if err != nil {
+		return tx, xerrors.Errorf("couldn't hash tx: %v", err)
+	}
+
+	tx.transaction = transaction{hash: hash}
+
+	return tx, nil
+}
+
+// FromProto implements ledger.TransactionFactory. It returns a new transaction
+// built from the protobuf message.
+func (f TransactionFactory) FromProto(pb proto.Message) (consumer.Transaction, error) {
+	var txProto proto.Message
+
+	switch in := pb.(type) {
+	case *any.Any:
+		var err error
+		txProto, err = f.encoder.UnmarshalDynamicAny(in)
+		if err != nil {
+			return nil, xerrors.Errorf("couldn't unmarshal input: %v", err)
+		}
+	default:
+		txProto = pb
+	}
+
+	switch tx := txProto.(type) {
+	case *SpawnTransactionProto:
+		arg, err := f.encoder.UnmarshalDynamicAny(tx.GetArgument())
+		if err != nil {
+			return nil, xerrors.Errorf("couldn't unmarshal argument: %v", err)
+		}
+
+		return f.NewSpawn(tx.GetContractID(), arg)
+	case *InvokeTransactionProto:
+		arg, err := f.encoder.UnmarshalDynamicAny(tx.GetArgument())
+		if err != nil {
+			return nil, xerrors.Errorf("couldn't unmarshal argument: %v", err)
+		}
+
+		return f.NewInvoke(tx.GetKey(), arg)
+	case *DeleteTransactionProto:
+		return f.NewDelete(tx.GetKey())
+	default:
+		return nil, xerrors.Errorf("invalid transaction type '%T'", txProto)
+	}
+}
+
+// ContractInstance is a specialization of the consumer instance to include
+// smart contract details.
+type ContractInstance interface {
+	consumer.Instance
+
+	GetContractID() string
+	Deleted() bool
+}
+
+// contractInstance is the result of a smart contract transaction execution. The
+// key is defined by the hash of the spawn transaction that created the
```

Comment:
How can you be sure that the key is uniq ?

Reply:
It is indeed assuming that the transaction hash is unique for every transaction which is expected from it.